python/cuspatial/cuspatial/geometry/geoseries.py

```diff
@@ -55,75 +64,249 @@ def __init__(
 
     @property
     def lines(self):
         """
         Access the `LineArray` of the underlying `GeoArrowBuffers`.
         """
-        return self._geocolumn.lines
+        return self.GeoColumnAccessor(self._column.lines)
 
     @property
     def polygons(self):
         """
         Access the `PolygonArray` of the underlying `GeoArrowBuffers`.
         """
-        return self._geocolumn.polygons
+        return self.GeoColumnAccessor(self._column.polygons)
 
     def __repr__(self):
         # TODO: Implement Iloc with slices so that we can use `Series.__repr__`
         return self.to_pandas().__repr__()
 
+    class GeoSeriesLocIndexer:
+        """
+        Not yet supported.
+        """
+
+        def __init__(self):
+            # Todo: Easy to implement with a join.
+            raise NotImplementedError
+
+    class GeoSeriesILocIndexer:
+
+        """
+        Each row of a GeoSeries is one of the six types: Point, MultiPoint,
+        LineString, MultiLineString, Polygon, or MultiPolygon.
+        """
+
+        def __init__(self, sr):
+            self._sr = sr
+
+        @cached_property
+        def _type_int_to_field(self):
+            return {
+                Feature_Enum.POINT: self._sr.points,
+                Feature_Enum.MULTIPOINT: self._sr.mpoints,
+                Feature_Enum.LINESTRING: self._sr.lines,
+                Feature_Enum.MULTILINESTRING: self._sr.lines,
+                Feature_Enum.POLYGON: self._sr.polygons,
+                Feature_Enum.MULTIPOLYGON: self._sr.polygons,
+            }
+
+        @cached_property
+        def _get_shapely_class_for_Feature_Enum(self):
+            type_map = {
+                Feature_Enum.POINT: Point,
+                Feature_Enum.MULTIPOINT: MultiPoint,
+                Feature_Enum.LINESTRING: LineString,
+                Feature_Enum.MULTILINESTRING: MultiLineString,
+                Feature_Enum.POLYGON: Polygon,
+                Feature_Enum.MULTIPOLYGON: MultiPolygon,
+            }
+            return type_map
+
+        def __getitem__(self, item):
+            """
+            Use a host copy of a pyarrow DenseUnion to index into
+            the GPU data for local copies.
+            """
+            # Use the reordering _column._data if it is set
+            index = (
+                self._sr._column._data[item].to_pandas()
+                if self._sr._column._data is not None
+                else item
+            )
+
+            # Fix types: There's only four fields
+            result_types = self._sr._column._meta.input_types.to_arrow()
+            union_types = self._sr._column._meta.input_types.replace(3, 2)
+            union_types = union_types.replace(4, 3)
+            union_types = union_types.replace(5, 3).values_host
+            shapely_classes = [
+                self._get_shapely_class_for_Feature_Enum[Feature_Enum(x)]
+                for x in result_types.to_numpy()
+            ]
```

Comment:
`shapely_classes` seems like a preprocess pass in an attempt to map the type code to its corresponding host type. I was expecting to see in the following code the iteration can be simplified without if-else branches, but apparently they still exists. Makes me wonder if it's better to just directly call the shapely class in the loop and avoid this overhead.

Reply:
:+1: Better might be just to put the logic for the serializing into functions that are identified here, so that the loop doesn't need a conditional. Good idea.

Reply:
Done.